### DISCUSSION TITLE
Duplicate uber-trace-id generated with query-frontend module

### DISCUSSION THREAD

**Discussion by mengmengy** [cortexproject/cortex#4643] [Q&A]
Hi All,

I created a QF service just wrapping a cortex instance and with only query-frontend module enabled. I didn't enable jaeger to send traces, but I do need some unique Id for each reqeust, so I added below code to my QF main.go. 
```
if trace, err := tracing.NewFromEnv(name); err != nil {
	level.Error(util_log.Logger).Log("msg", "Failed to setup tracing", "err", err.Error())
} else {
	defer trace.Close()
}
```
Then the requests coming from my QF service will have a uber-trace-id(genereated in cortex query-frontend module) in the header to my query service. I'm expecting the uber-trace-id is unique, however, I'm keep seeing different client requests having the same uber-trace-id and the reqeusts were made days apart. Below are example requests having same uber-trace-id
```
e.g. 
request 1: 2022-01-21T06:17:27.00316Z POST /api/v1/query 03bd628ecdcd5232:03bd628ecdcd5232:0000000000000000:0
request 2: 2022-01-18T18:10:27.548208Z GET /api/v1/query 03bd628ecdcd5232:03bd628ecdcd5232:0000000000000000:0
```
I assume duplicate uber-trace-id is not expected, did I miss anything to generate uber-trace-id? Could you please help to find the root cause of this?

**Comment by bboreham** [comment]
I looked at the code and it's randomly generated; if you do enough of them you would get a clash.
Since the top-level span ID and subsequent trace IDs are all generated from the same (pool of) generators I guess it chews through them a bit faster than if there as just one generator for trace IDs.

You could try setting env var `JAEGER_TRACEID_128BIT`.

**Reply by mengmengy** [reply]
> I looked at the code and it's randomly generated; if you do enough of them you would get a clash. Since the top-level span ID and subsequent trace IDs are all generated from the same (pool of) generators I guess it chews through them a bit faster than if there as just one generator for trace IDs.
> 
> You could try setting env var `JAEGER_TRACEID_128BIT`.

I see, thank you helping look into this. Since jaeger use math/rand to generate a random number with [time.Now().UnixNano()](https://github.com/jaegertracing/jaeger-client-go/blob/master/tracer.go#L126) as seed, which means the random number is predicable for each run. Setting env `JAEGER_TRACEID_128BIT` can generate additional High bits for traceId, but this is not guranteed to make uber-trace-id unique, right?
